crypto/ec/asm/ecp_nistz256-x86_64.pl

```diff
@@ -3618,17 +3618,26 @@ ()
 
 	or	$acc5, $acc4			# see if result is zero
 	or	$acc0, $acc4
-	or	$acc1, $acc4
+	or	$acc1, $acc4			# !is_equal(U1, U2)
 
-	.byte	0x3e				# predict taken
-	jnz	.Ladd_proceed$x			# is_equal(U1,U2)?
 	movq	%xmm2, $acc0
 	movq	%xmm3, $acc1
-	test	$acc0, $acc0
-	jnz	.Ladd_proceed$x			# (in1infty || in2infty)?
+
+	or	$acc0, $acc4
+	.byte	0x3e				# predict taken
+	jnz	.Ladd_proceed$x			# !is_equal(U1, U2) || in1infty || in2infty
```

Comment:
@davidben, would you mind explaining this?  Before the change, `or $acc1, $acc4` is the instruction that affects the following `jnz`, but here it's `or $acc0, $acc4`, and yet, I can't quite see how `$acc0` and `$acc1` would represent the same value, or how an `or` would affect ZF the same way `test` does, or how `or $acc0, $acc4` alone can test so many things at once...

Reply:
> Before the change, or $acc1, $acc4 is the instruction that affects the following jnz, but here it's or $acc0, $acc4, and yet, I can't quite see how $acc0 and $acc1 would represent the same value

They're not the same value. That's the point. The old OpenSSL branching leaked secret information.

>  or how an or would affect ZF the same way test does

In both cases, it's whether the computed value is zero. Hence *zero* flag.

> or how or $acc0, $acc4 alone can test so many things at once...

Recall that x86 instructions write to registers. :-P

Lines 3619 through 3621 set $acc4 to $acc0 | $acc1 | $acc4 | $acc5 and ZF based on whether $acc4 is now zero. Thus lines 3619 through 3628 in the old code is:

```
$acc4 = $acc0 | $acc1 | $acc4 | $acc5
if ($acc4 != 0) {
   goto add_proceed
}
$acc0 = %xmm2
$acc1 = %xmm3
if ($acc0 != 0) {
   goto add_proceed
}
[more stuff]
```

This leaks whether we end to add_proceed based on the first check (do x coordinates of the points match, i.e. P = +/- Q) or the second check (whether P is infinity). If you look at the point multiplication setup, this is not okay and leaks secret information. The new version is:

```
$acc4 = $acc0 | $acc1 | $acc4 | $acc5
$acc0 = %xmm2
$acc1 = %xmm3
$acc4 |= $acc0
if ($acc4 != 0) {
   goto add_proceed
}
[more stuff]
```

This is equivalent because X != 0 || Y != 0 is the same as (X|Y) != 0, and because add_proceed does not care about $acc0 and $acc1. $acc1 is loaded for the benefit of the "[more stuff]" bits, which check for P + (-P).

Note this is not the end of the review story because this code still branches. To check whether the remaining branches are safe, the discussion in BoringSSL here is a useful framework:
https://boringssl.googlesource.com/boringssl/+/refs/heads/master/crypto/fipsmodule/ec/util.c
Note, however, the full conclusions only apply to one particular addition schedule and do not cover the P + (-P). I've done the rest of that analysis and I believe it's okay, though I still need to generalize some of the theorems in that comment.

Reply:
> Recall that x86 instructions write to registers. :-P

Oh, of course!  How could I forget...

Reply:
... and with that reminder, the change makes much more sense, thanks (and thanks for making it easy on me with the code examples)